test/bio_base64_test.c

```diff
@@ -0,0 +1,172 @@
+    { z256enc softEOF, 265, zbin, 192, 0, "256 encoded bytes + soft EOF" },
+    { z256enc junk, 265, zbin, 0, -1, "256 encoded bytes + junk" },
+    { z1024enc, 1040, zbin, 768, 0, "1024 encoded bytes" },
+    { junk z1024enc, 1045, zbin, 768, 0, "junk + 1024 encoded bytes" },
+    { z1024enc softEOF, 1045, zbin, 768, 0, "1024 encoded bytes + soft EOF" },
+    { z1024enc junk, 1045, zbin, 720, -1, "1024 encoded bytes + junk" },
+    { z2048enc, 2080, zbin, 1536, 0, "2048 encoded bytes" },
+    { junk z2048enc, 2085, zbin, 1536, 0, "junk + 2048 encoded bytes" },
+    { z2048enc softEOF, 2085, zbin, 1536, 0, "2048 encoded bytes + softEOF" },
+    { z2048enc junk, 2085, zbin, 1488, -1, "2048 encoded bytes + junk" },
+    { gunk z2048enc, 3105, zbin, 1536, 0, "gunk + 2048 encoded bytes" },
+    { NULL, 0, NULL, 0, 0, NULL }
+};
+
+static int single_test(test_case *t, int eof_return)
+{
+    char *out;
+    BIO *bio, *b64;
+    int out_len;
+    int n, n1, n2;
+    int ret;
+
+    bio = BIO_new(BIO_s_mem());
+    if (eof_return <= 0)
+        BIO_set_mem_eof_return(bio, eof_return);
+    else
+        eof_return = 0;
+
+    /*
+     * When the input is long enough, and the source bio is retriable, test
+     * retries by writting it two steps (1024 bytes, then the rest).
+     */
+    n1 = t->encoded_len;
+    if (eof_return < 0 && n1 > 1024)
+        n1 = 1024;
+    if (n1 > 0)
+        BIO_write(bio, t->encoded, n1);
+
+    b64 = BIO_new(BIO_f_base64());
+    BIO_push(b64, bio);
+
+    out_len = t->expected_decode_len + 1024;
+    out = OPENSSL_malloc(out_len);
+    n = BIO_read(b64, out, out_len);
+
+    /* Retry when we have more input */
+    if (n1 < t->encoded_len) {
+        /* Append the rest of the input, and read again */
+        BIO_write(bio, t->encoded + n1, t->encoded_len - n1);
+        if (n > 0) {
+            n2 = BIO_read(b64, out + n, out_len - n);
+            if (n2 > 0)
+                n += n2;
+        } else if (n == eof_return) {
+            n = BIO_read(b64, out, out_len);
+        }
+    }
+
+    /* Turn retry-related negative results to normal (0) EOF */
+    if (n < 0 && n == eof_return)
+        n = 0;
+
+    /* Turn off retries */
+    if (eof_return < 0)
+        BIO_set_mem_eof_return(bio, 0);
+
+    if (n < out_len) {
+        /* Perform the last read, checking its result */
+        ret = BIO_read(b64, out + n, out_len - n);
+    } else {
+        /* Should not happen, given extra space in out_len */
+        ret = t->last_read_ret - 1;
+    }
+
+    /* Should now equal the expected last read return */
+    ret = ret == t->last_read_ret ? 0 : -1;
+
+    /* After decoding the expected output, if any */
+    if (n == t->expected_decode_len) {
+        if (memcmp(t->expected_decode, out, n) != 0)
+            ret = -1;
+    } else if (t->expected_decode_len != 0 || t->last_read_ret == 0) {
+        ret = -1;
+    }
+
+    BIO_free_all(b64);
+    OPENSSL_free(out);
+
+    return ret;
+}
+
+int main(void)
```

Comment:
Just to ask the question, why not use the test framework here with the ADD_ALL_TESTS macro?  Particularly since you created the dependency on libtestutil.a in the build.info file below?

Reply:
I am not yet sufficiently familiar with the test framework, and just imported the RNG for now.
Given the initial tests, it was easier to stick with the naïve approach. Perhaps next time around I'll find the cycles to learn about the rest of the test framework.

Reply:
That's just an FYI comment, right? Nothing for me to do at present?
The  machinery would be good to eventually be comfortable with, but if I understood you correctly, yes is presently sufficiently compelling to immediately rework the test harness in this PR.

Reply:
yeah, just wanted to know why you didn't use the existing test harness, like perhaps there may have been a motivating reason.  May be worth converting in the future, not tied to it at the moment though

Reply:
Not sure if not using the proper TEST macros and the test harness in the test should be a blocker or not. What are the @mattcaswell @levitte @paulidale @openssl/otc opinions?